#!/bin/sh

#bowtie -f  --phred64-quals  -n 1 -e 3440 -l 25 -y  -k 10 --nomaqround -p 6 --sam  hg19/hg19 /home/zhengqi/git/AlignerBoost/AlignerBoost/test/human_test1_NR.fas | samtools view -S -b -o /home/zhengqi/git/AlignerBoost/AlignerBoost/test/human_test1_hg19_raw.bam -
#bowtie2 -q  --local  --phred64  -N 1 -L 25 -k 10 -p 6  -x hg19/hg19 -U /home/zhengqi/git/AlignerBoost/AlignerBoost/test//home/zhengqi/git/AlignerBoost/AlignerBoost/test/human_test2_trimmed.fastq | samtools view -S -b -o /home/zhengqi/git/AlignerBoost/AlignerBoost/test/human_test2_hg19_raw.bam -
#bwa mem -t 6 -k 25 -T 15  -a hg19/hg19 /home/zhengqi/git/AlignerBoost/AlignerBoost/test//home/zhengqi/git/AlignerBoost/AlignerBoost/test/human_test3_trimmed.fastq | samtools view -S -b -o /home/zhengqi/git/AlignerBoost/AlignerBoost/test/human_test3_hg19_raw.bam -
#bwa bwasw -t 6 -T 13 -z 10  hg19/hg19 /home/zhengqi/git/AlignerBoost/AlignerBoost/test//home/zhengqi/git/AlignerBoost/AlignerBoost/test/human_test4_trimmed.fastq | samtools view -S -b -o /home/zhengqi/git/AlignerBoost/AlignerBoost/test/human_test4_hg19_raw.bam -
#bwa aln -n 0.06 -l 25 -k 1 -t 6 -I  hg19/hg19 /home/zhengqi/git/AlignerBoost/AlignerBoost/test//home/zhengqi/git/AlignerBoost/AlignerBoost/test/human_test5_trimmed.fastq > human_test5_hg19.sai
#  bwa samse hg19/hg19 human_test5_hg19.sai /home/zhengqi/git/AlignerBoost/AlignerBoost/test//home/zhengqi/git/AlignerBoost/AlignerBoost/test/human_test5_trimmed.fastq | samtools view -S -b -o /home/zhengqi/git/AlignerBoost/AlignerBoost/test/human_test5_hg19_raw.bam -
#tophat2 --bowtie1 -N 6 --read-gap-length 0 --read-edit-dist 6 -g 10 -x 10 --max-insertion-length 0 --max-deletion-length 0 --solexa1.3-quals --library-type fr-firststrand -p 6 -G hg19/refGene.gff --transcriptome-index hg19/tophat1/refGene --segment-length 25 --segment-mismatches 1  --no-sort-bam  -o human_test6_hg19_tophat1 hg19/hg19 /home/zhengqi/git/AlignerBoost/AlignerBoost/test//home/zhengqi/git/AlignerBoost/AlignerBoost/test/human_test6_trimmed.fastq
#mv human_test6_hg19_tophat1/accepted_hits.bam /home/zhengqi/git/AlignerBoost/AlignerBoost/test/human_test6_hg19_raw.bam
#tophat2 -N 6 --read-gap-length 0 --read-edit-dist 6 -g 10 -x 10 --max-insertion-length 0 --max-deletion-length 0 --solexa1.3-quals --library-type fr-firststrand -p 6 -G hg19/refGene.gff --transcriptome-index hg19/tophat2/refGene --segment-length 25 --segment-mismatches 1--b2-N 1--b2-L 25  --no-sort-bam  -o human_test7_hg19_tophat2 hg19/hg19 /home/zhengqi/git/AlignerBoost/AlignerBoost/test//home/zhengqi/git/AlignerBoost/AlignerBoost/test/human_test7_trimmed.fastq
#mv human_test7_hg19_tophat2/accepted_hits.bam /home/zhengqi/git/AlignerBoost/AlignerBoost/test/human_test7_hg19_raw.bam
#STAR --genomeDir hg19/STAR --readFilesIn /home/zhengqi/git/AlignerBoost/AlignerBoost/test//home/zhengqi/git/AlignerBoost/AlignerBoost/test/human_test8_trimmed.fastq --runThreadN 6 --outFilterScoreMin 0.94 --outFilterMatchNminOverLread 0.94 --outFilterMultimapScoreRange 4 --outFilterMultimapNmax 10 --outFilterMismatchNmax 6 --outFilterMismatchNoverLmax 0.06  --outStd SAM - | samtools view -S -b -o /home/zhengqi/git/AlignerBoost/AlignerBoost/test/human_test8_hg19_raw.bam -
